src/extensions/arith.rs

```diff
@@ -0,0 +1,1080 @@
+            ),
+            ExprInner::Output(i) => (
+                script_num_size(*i) + 3, // i INPSECTOUTPUTVALUE <1> EQUALVERIFY
+                0,
+            ),
+            ExprInner::InputIssue(i) => (
+                script_num_size(*i) + 7, // i OP_INSPECTINPUTISSUANCE DROP DROP <1> EQUALVERIFY NIP NIP
+                0,
+            ),
+            ExprInner::InputReIssue(i) => (
+                script_num_size(*i) + 7, // i OP_INSPECTINPUTISSUANCE DROP DROP DROP DROP <1> EQUALVERIFY
+                0,
+            ),
+            ExprInner::Add(x, y) => (
+                x.script_size + y.script_size + 3, // [X] [Y] ADD64 <1> EQUALVERIFY
+                cmp::max(x.depth, y.depth),
+            ),
+            ExprInner::Sub(x, y) => (
+                x.script_size + y.script_size + 3, // [X] [Y] SUB64 <1> EQUALVERIFY
+                cmp::max(x.depth, y.depth),
+            ),
+            ExprInner::Mul(x, y) => (
+                x.script_size + y.script_size + 3, // [X] [Y] MUL64 <1> EQUALVERIFY
+                cmp::max(x.depth, y.depth),
+            ),
+            ExprInner::Div(x, y) => (
+                x.script_size + y.script_size + 4, // [X] [Y] DIV64 <1> EQUALVERIFY NIP
+                cmp::max(x.depth, y.depth),
+            ),
+            ExprInner::Mod(x, y) => (
+                x.script_size + y.script_size + 4, // [X] [Y] DIV64 <1> EQUALVERIFY DROP
+                cmp::max(x.depth, y.depth),
+            ),
+            ExprInner::BitAnd(x, y) => (
+                x.script_size + y.script_size + 1, // [X] [Y] AND
+                cmp::max(x.depth, y.depth),
+            ),
+            ExprInner::BitOr(x, y) => (
+                x.script_size + y.script_size + 1, // [X] [Y] OR
+                cmp::max(x.depth, y.depth),
+            ),
+            ExprInner::Xor(x, y) => (
+                x.script_size + y.script_size + 1, // [X] [Y] XOR
+                cmp::max(x.depth, y.depth),
+            ),
+            ExprInner::Invert(x) => (
+                x.script_size + 1, // [X] INVERT
+                x.depth + 1,
+            ),
+            ExprInner::Negate(x) => (
+                x.script_size + 3, // [X] NEG64 <1> EQUALVERIFY
+                x.depth + 1,
+            ),
+        };
+        Self {
+            inner,
+            script_size,
+            depth,
+        }
+    }
+
+    /// Evaluate this expression
+    fn eval(
+        &self,
+        tx: &elements::Transaction,
+        utxos: &[elements::TxOut],
+    ) -> Result<i64, EvalError> {
+        match &self.inner {
+            ExprInner::Const(c) => Ok(*c),
+            ExprInner::Input(i) => {
+                if *i >= utxos.len() {
+                    return Err(EvalError::UtxoIndexOutOfBounds(*i, utxos.len()));
+                }
+                utxos[*i]
+                    .value
+                    .explicit()
+                    .map(|x| x as i64) // safe conversion bitcoin values from u64 to i64 because 21 mil
```

Comment:
In 97f16a908cb974e4001308a2418cef6e32130385:

In theory, in Elements we could have some asset with more than 2^63 tokens issued. This will probably break a number of wallet related things, so I don't think that the evaluator logic here is the highest priority to spend time working out the consequences, but it's worth thinking about.